docs/architecture-decision-records/20220121-conduit-plugin-architecture.md

```diff
@@ -0,0 +1,322 @@
+      call `Ack` which will block until either an error occurs or an acknowledgment is produced by the plugin.
+
+### Plugin registries
+
+We introduced a plugin dispenser interface with two implementations that allow us to interact with the plugin either
+through go-plugin (i.e. plugin runs in a standalone process) or directly through Go (i.e. plugin acts as a library).
+
+```go
+type Dispenser interface {
+  DispenseSpecifier() (SpecifierPlugin, error)
+  DispenseSource() (SourcePlugin, error)
+  DispenseDestination() (DestinationPlugin, error)
+}
+```
+
+The dispenser is responsible for managing the plugin lifecycle, from spinning up the plugin (i.e. starting the process)
+to tearing it down (i.e. killing the process), depending on if there is any dispensed plugin left. There are two
+implementations of a dispenser:
+
+- The built-in dispenser will dispense plugins that communicate with the plugin via direct Go calls. This means that
+  Conduit can import a plugin as any other library, register it in the built-in plugin registry and use it without
+  having a separate binary for the plugin. The implementation behind the built-in plugin is using the thin layer defined
+  in the conduit-plugin repository (see previous chapter), meaning that it is irrelevant which SDK the plugin uses (if
+  any), as long as it implements the interfaces defined in that thin layer.
+- The standalone dispenser will dispense plugins that communicate with the plugin via go-plugin and gRPC. It will start
+  the plugin in a separate process as soon as a call to dispense a plugin comes in. When the method `Teardown` gets
+  called the dispenser regards the plugin as done and the plugin process will be stopped. If then a new plugin tries to
+  be dispensed the process will be started again. Note that the specifier plugin does not contain a `Teardown` method so
+  a call to `Specify` will signal to the dispenser that the plugin process can be stopped.
+
+The plugin registry is the last missing piece that contains information about the available plugins.
+
+```protobuf
+type registry interface {
+  New(logger log.CtxLogger, name string) (Dispenser, error)
+    }
+```
+
+Behind this interface, there are again two registries that contain either built-in or standalone dispensers. To
+differentiate which registry should be used we use a prefix in the name of the plugin - any plugin that starts
+with `builtin:` will be taken from the built-in registry. Note that the naming of the prefix is not final and is subject
+to change.
+
+### Conduit plugin SDK
+
+The Conduit plugin SDK is currently a package in Conduit but is written as a standalone package that does not depend on
+any internal structs from Conduit. This will aid in extracting the SDK into a separate repository in the future.
+
+The goal of the SDK package is to hide the complexity and verbosity of the plugin interface and provide utilities for
```

Comment:
Here I was thinking about the plugin interface defined in [conduit-plugin](https://github.com/conduitio/conduit-plugin). When I say "verbosity" I mean the fact that every endpoint expects a message, when it could directly expect the parameters it needs. This is done so that the proto interface is extensible, but the SDK hides the verbosity by transforming the message into minimal structs that are closer to what a developer is used to work with. And when I say "complexity" I'm mostly talking about the fact that we have a bidirectional stream which means that the plugin has to run 2 goroutines, one for receiving messages and one for sending them. This is also hidden in the SDK from the developer, they don't have to care about managing these goroutines.

Reply:
Sweet, thanks for the detailed explanation. 👍